### PR TITLE
Add validate_email_address function

### PR DESCRIPTION
**Pull request by jyaworski**

### CODE REVIEW

**Comment by jyaworski** [comment]
@hunner @bmjen what's this failure? It passes locally for me.

**Comment by bmjen** [comment]
@jyaworski 
```
Failures:
  1) get_module_path should run get_module_path("one") and raise an Puppet::Environments::EnvironmentNotFound with the message matching /Could not find a directory environment/
     Failure/Error: it { is_expected.to run.with_params('one').and_raise_error(Puppet::Environments::EnvironmentNotFound, /Could not find a directory environment/) }
       expected get_module_path("one") to have raised Puppet::Environments::EnvironmentNotFound matching /Could not find a directory environment/ instead of Puppet::ParseError(Could not find module one in environment rp_env)
     # ./spec/functions/get_module_path_spec.rb:9:in `block (2 levels) in <top (required)>'
```

Looks like the output changed in latest Puppet 4.x?

**Comment by bmjen** [comment]
Doesn't look related to your PR:

https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/spec/functions/get_module_path_spec.rb#L8-L12

**Comment by jyaworski** [comment]
Let me know when I should rebase.

File: lib/puppet/parser/functions/is_email_address.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
...[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*)(?:\.(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*))*(?:,@(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*)(?:\.(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*))*)*:(?:(?:\r\n)?[ \t])*)?(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|"(?:[^\"\r\\]|\\.|(?:(?:\r\n)?[ \t]))*"(?:(?:\r\n)?[ \t])*)(?:\.(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|"(?:[^\"\r\\]|\\.|(?:(?:\r\n)?[ \t]))*"(?:(?:\r\n)?[ \t])*))*@(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*)(?:\.(?:(?:\r\n)?[ \t])*(?:[^()<>@,;:\\".\[\] \000-\031]+(?:(?:(?:\r\n)?[ \t])+|\Z|(?=[\["()<>@,;:\\".\[\]]))|\[([^\[\]\r\\]|\\.)*\](?:(?:\r\n)?[ \t])*))*\>(?:(?:\r\n)?[ \t])*))*)?;\s*)/
```
Comment (igalic) [inline lib/puppet/parser/functions/is_email_address.rb]: can't we use `/x` here to make this readable?

Comment (jyaworski) [inline lib/puppet/parser/functions/is_email_address.rb]: I _guess_ I can do that.

Comment (igalic) [inline lib/puppet/parser/functions/is_email_address.rb]: please, for the love of humanity, do that.

Comment (jyaworski) [inline lib/puppet/parser/functions/is_email_address.rb]: I'm actually not sure that I can.

There are whitespaces in the actual expression.

**Comment by jyaworski** [comment]
I tried using heredoc, but it looks like it's unable to distinguish between literal \n in the string and the multi-line patterns. Maybe we should accept it as a single-line string @bmjen. It's not pretty, but it works.

**Comment by tphoney** [comment]
@jyaworski thanks for making the change, it looks like the CI is failing. Can you have a look ?

**Comment by jyaworski** [comment]
@tphoney it's failing due to recommendations by @igalic and I wanted some clarification before I fixed it. Are we fine with one absolutely gigantic line? Other ways seem to fail.

**Comment by jyaworski** [comment]
@tphoney we're good now.

**Comment by tphoney** [comment]
LGTM !!!! @jyaworski thanks

**Pull request merged**